### PR TITLE
"Best practices" for more internal API

### PR DESCRIPTION
**Pull request by kgaillot**
Most of this is follow-up to earlier parts of this project, covering bits and pieces that were missed. There are also some logging improvements and one very minor memory leak fix found during refactoring.

### CODE REVIEW

**Comment by kgaillot** [comment]
@clumens , low priority, but when you get a chance can you review this?

File: include/crm/common/acl.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#  include <libxml/tree.h> // xmlNode
+
+bool xml_acl_enabled(xmlNode *xml);
+void xml_acl_disable(xmlNode *xml);
+bool xml_acl_denied(xmlNode *xml);
+bool xml_acl_filtered_copy(const char *user, xmlNode* acl_source, xmlNode *xml,
+                           xmlNode **result);
+
+bool pcmk_acl_required(const char *user);
+
```
Comment (clumens) [inline include/crm/common/acl.h]: Public API should also get documentation, right?

File: include/crm/common/internal.h [hunk 2]
```diff
@@ -162,13 +162,6 @@ int pcmk__compress(const char *data, unsigned int length, unsigned int max,
 // Example: crm_info("Found %d node%s", nnodes, pcmk__plural_s(nnodes));
 #define pcmk__plural_s(i) pcmk__plural_alt(i, "", "s")
 
-static inline char *
-crm_concat(const char *prefix, const char *suffix, char join)
-{
-    CRM_ASSERT(prefix && suffix);
-    return crm_strdup_printf("%s%c%s", prefix, join, suffix);
-}
-
```
Comment (clumens) [inline include/crm/common/internal.h]: This patch makes things so much easier to read.  The argument order to crm_concat was really awkward.

File: daemons/controld/controld_fencing.c [hunk 3]
```diff
@@ -847,11 +848,11 @@ te_fence_node(crm_graph_t *graph, crm_action_t *action)
                                   (int) (transition_graph->stonith_timeout / 1000),
                                   0);
 
+    transition_key = pcmk__transition_key(transition_graph->id, action->id, 0,
+                                          te_uuid),
     stonith_api->cmds->register_callback(stonith_api, rc,
                                          (int) (transition_graph->stonith_timeout / 1000),
-                                         st_opt_timeout_updates,
-                                         generate_transition_key(transition_graph->id, action->id,
-                                                                 0, te_uuid),
```
Comment (clumens) [inline daemons/controld/controld_fencing.c]: I haven't dug into it further, but do the results of pcmk__transition_key need to be freed?  The fact that it's calling crm_strdup_printf would make me think yes, but I don't know what the callback is going to do with the data.  There are several other places in this patch where the same question would apply.

File: include/crm/common/acl.h (hunk 1 above)
Comment (kgaillot) [reply include/crm/common/acl.h]: Good point :)

File: include/crm/common/internal.h (hunk 2 above)
Comment (kgaillot) [reply include/crm/common/internal.h]: Yeah that was my favorite too :)

File: daemons/controld/controld_fencing.c (hunk 3 above)
Comment (kgaillot) [reply daemons/controld/controld_fencing.c]: Good eye. In these cases that's the user data passed to the callback, and the callback frees it. I'm not thrilled that it's non-obvious but it's correct.

File: include/crm/common/acl.h (hunk 1 above)
Comment (kgaillot) [reply include/crm/common/acl.h]: Good point, updated

**Comment by kgaillot** [comment]
Only Debian 10 failed CI and that was a jenkins error, this should be good

**Pull request merged**